### PR TITLE
 Transactional API for the Trie database back-end

### PR DESCRIPTION
**Pull request by zah**
See the commit messages for more details.

### CODE REVIEW

File: eth_trie/db.nim [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import types
+export types
+
```
Comment (zah) [inline eth_trie/db.nim]: The contents of the types file will be moved here eventually (after getting rid of `eth_trie/types` imports in other projects).

File: eth_trie/memdb.nim [hunk 2]
```diff
@@ -1,71 +1,3 @@
...
-  result.new
-  result.tbl = initTable[Bytes, MemDbRecord]()
-  put(result, emptyRlpHash.data, emptyRlp.toOpenArray)
-
-proc len*(db: MemDB): int =
-  db.tbl.len
-
-#static:
-#  assert MemDB is TrieDatabaseConcept
+import types
```
Comment (zah) [inline eth_trie/memdb.nim]: This module will be removed (after getting rid of its imports in other projects)

File: eth_trie/types.nim [hunk 3]
```diff
@@ -20,12 +34,128 @@ type
...
+    db.del(k)
+
+  for k, v in memDb.records:
+    db.put(k, v.value)
+
+proc init(db: var MemDB) =
+  db = newMemDB()
+
+proc newMemoryDB*: TrieDatabaseRef =
+  new result
```
Comment (zah) [inline eth_trie/types.nim]: This is a bit tricky. The new way to create in-memory databases is to just start a transaction that will never be rolled back or committed. We'll be able to use the same trick for the transient databases required for stateless clients.

**Comment by zah** [comment]
@yglukhov, it hides it a bit more under the rug, although I still plan to look deeper into the issue. I noticed that the C++ client processes the entire blocks in memory before committing them (it terms of this API, it creates a transaction). Besides mitigating the ref counting problem to some extent, this may also be a performance optimization, because the transaction processing is probably generating a lot of temporary state that will be discarded before being written to disk.

**Comment by coffeepots** [comment]
Looks good!

**Pull request merged**